pennylane_lightning/lightning_gpu/_state_vector.py

```diff
@@ -0,0 +1,338 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Class implementation for lightning_gpu state-vector manipulation.
+"""
+from warnings import warn
+
+try:
+    from pennylane_lightning.lightning_gpu_ops import StateVectorC64, StateVectorC128
+
+    try:  # Try to import the MPI modules
+        from pennylane_lightning.lightning_gpu_ops import StateVectorMPIC64, StateVectorMPIC128
+
+        mpi_error = None
+        MPI_SUPPORT = True
+    except ImportError as ex_mpi:
+        mpi_error = ex_mpi
+        MPI_SUPPORT = False
+
+except ImportError as ex:
+    warn(str(ex), UserWarning)
+
+from itertools import product
+from typing import Union
+
+import numpy as np
+import pennylane as qml
+from pennylane import DeviceError
+from pennylane.ops.op_math import Adjoint
+from pennylane.wires import Wires
+
+# pylint: disable=ungrouped-imports
+from pennylane_lightning.core._serialize import global_phase_diagonal
+from pennylane_lightning.core._state_vector_base import LightningBaseStateVector
+
+from ._mpi_handler import MPIHandler
+
+gate_cache_needs_hash = (
+    qml.BlockEncode,
+    qml.ControlledQubitUnitary,
+    qml.DiagonalQubitUnitary,
+    qml.MultiControlledX,
+    qml.OrbitalRotation,
+    qml.PSWAP,
+    qml.QubitUnitary,
+)
+
+
+class LightningGPUStateVector(LightningBaseStateVector):
+    """Lightning GPU state-vector class.
+
+    Interfaces with C++ python binding methods for state-vector manipulation.
+
+    Args:
+        num_wires(int): the number of wires to initialize the device with
+        dtype: Datatypes for state-vector representation. Must be one of
+            ``np.complex64`` or ``np.complex128``. Default is ``np.complex128``
+        device_name(string): state vector device name. Options: ["lightning.gpu"]
+        mpi_handler(MPIHandler): MPI handler for PennyLane Lightning GPU device.
+            Provides functionality to distribute the state-vector to multiple devices.
+        sync (bool): immediately sync with host-sv after applying operation.
```

Comment:
It will be more accurate to describe it as: is host-device data copy synchronized or not

Reply:
I will change it